x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/date/DateDiff.java

```diff
@@ -0,0 +1,216 @@
+            VALID_VALUES = DateTimeField.initializeValidValues(values());
+        }
+
+        private BiFunction<ZonedDateTime, ZonedDateTime, Integer> diffFunction;
+        private Set<String> aliases;
+
+        Part(BiFunction<ZonedDateTime, ZonedDateTime, Integer> diffFunction, String... aliases) {
+            this.diffFunction = diffFunction;
+            this.aliases = Set.of(aliases);
+        }
+
+        public Integer diff(ZonedDateTime startTimestamp, ZonedDateTime endTimestamp) {
+            return diffFunction.apply(startTimestamp, endTimestamp);
+        }
+
+        @Override
+        public Iterable<String> aliases() {
+            return aliases;
+        }
+
+        private static int safeInt(long diff) {
+            if (diff > Integer.MAX_VALUE || diff < Integer.MIN_VALUE) {
+                throw new InvalidArgumentException(
+                    "The DATE_DIFF function resulted in an overflow; the number of units "
+                        + "separating two date/datetime instances is too large. Try to use DATE_DIFF with a less precise unit."
+                );
+            } else {
+                return Long.valueOf(diff).intValue();
+            }
+        }
+
+        public static List<String> findSimilar(String match) {
+            return DateTimeField.findSimilar(NAME_TO_PART.keySet(), match);
+        }
+
+        public static Part resolve(String dateTimeUnit) {
+            return DateTimeField.resolveMatch(NAME_TO_PART, dateTimeUnit);
+        }
+    }
+
+    @FunctionInfo(returnType = "integer",
+        description = "Subtract 2 dates and return their difference in multiples of a unit specified in the 1st argument")
+    public DateDiff(
+        Source source,
+        @Param(name = "unit", type = { "text" }, description = "A valid date unit") Expression unit,
+        @Param(
+            name = "startTimestamp",
+            type = { "date" },
+            description = "A string representing a start timestamp"
+        ) Expression startTimestamp,
+        @Param(name = "endTimestamp", type = { "date" }, description = "A string representing a end timestamp") Expression endTimestamp
+    ) {
+        super(source, List.of(unit, startTimestamp, endTimestamp));
+        this.unit = unit;
+        this.startTimestamp = startTimestamp;
+        this.endTimestamp = endTimestamp;
+    }
+
+    @Evaluator
+    static int process(BytesRef unit, long startTimestamp, long endTimestamp) throws IllegalArgumentException {
+        ZonedDateTime zdtStart = ZonedDateTime.ofInstant(Instant.ofEpochMilli(startTimestamp), UTC);
+        ZonedDateTime zdtEnd = ZonedDateTime.ofInstant(Instant.ofEpochMilli(endTimestamp), UTC);
+
+        Part datePartField = Part.resolve(unit.utf8ToString());
+        if (datePartField == null) {
+            List<String> similar = Part.findSimilar(unit.utf8ToString());
+            if (similar.isEmpty()) {
+                throw new InvalidArgumentException(
```

Comment:
A nit on style: you could prepare the message in the two branches and throw just once.

Reply:
I agree, I have updated the code.